### PR TITLE
feat(BFormDatalist): Initial implementation

### PR DESCRIPTION
**Pull request by dwgray**
# Describe the PR

A clear and concise description of what the pull request does.

## Small replication

A small replication or video walkthrough can help demonstrate the changes made. This is optional, but can help observe the intended changes. A mentioned issue that contains a replication also works.

## PR checklist

**What kind of change does this PR introduce?** (check at least one)

- [ ] Bugfix :bug: - `fix(...)`
- [ ] Feature - `feat(...)`
- [ ] ARIA accessibility - `fix(...)`
- [ ] Documentation update - `docs(...)`
- [ ] Other (please describe)

**The PR fulfills these requirements:**

- [ ] Pull request title and all commits follow the [**Conventional Commits**](https://www.conventionalcommits.org/) convention or has an [**override**](https://github.com/googleapis/release-please#how-can-i-fix-release-notes) in this pull request body **This is very important, as the `CHANGELOG` is generated from these messages, and determines the next version type. Pull requests that do not follow conventional commits or do not have an override will be denied**

### CODE REVIEW

File: packages/bootstrap-vue-next/src/composables/useFormSelect.ts [hunk 1]
```diff
@@ -25,13 +25,14 @@ export default (options: MaybeRefOrGetter, props: Record<string, unknown>) => {
     const html: string = get(option, props.htmlField as string)
     const disabled: boolean = get(option, props.disabledField as string)
 
-    const opts: undefined | (unknown | Record<string, unknown>)[] = props.optionsField
-      ? get(option, props.optionsField as string)
-      : undefined
+    const opts: undefined | (unknown | Record<string, unknown>)[] = get(
```
Comment (dwgray) [inline packages/bootstrap-vue-next/src/composables/useFormSelect.ts:28]: I'm using this composable in Datalists, which don't support options groups, so they don't have the optionsField property.  There are probably other ways to get around this, but it seems like this is a more robust way to handle the default case for other uses of this composable anyway, unless I'm missing something?

File: packages/bootstrap-vue-next/src/composables/useFormSelect.ts [hunk 2]
```diff
@@ -25,13 +25,14 @@ export default (options: MaybeRefOrGetter, props: Record<string, unknown>) => {
...
+      option,
+      (props.optionsField as string) ?? 'options'
+    )
 
     if (opts !== undefined) {
       return {
-        label: get(option, props.labelField as string) || text,
+        label: get(option, (props.labelField as string) ?? 'label') || text,
```
Comment (dwgray) [inline packages/bootstrap-vue-next/src/composables/useFormSelect.ts:35]: same comment as with optionsField

File: packages/bootstrap-vue-next/src/components/BForm/BFormDatalist.vue [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+const computedId = useId(() => props.id, 'datalist')
+
+const {normalizedOptions, isComplex} = useFormSelect(() => props.options, props)
+
+const normalizedOptsWrapper = computed(() => {
+  let opts = normalizedOptions.value
+  if (opts.find((opt) => isComplex(opt))) {
+    // eslint-disable-next-line no-console
+    console.warn('Datalist may not contain complex options', computedId.value)
```
Comment (dwgray) [inline packages/bootstrap-vue-next/src/components/BForm/BFormDatalist.vue:50]: I'd like to use the same mechansisms we have in place for options lists, but datafields don't allow for options groups. I'm not seeing a consistent way to manage runtime errors for a case like this, happy with any guidance for managing this differently.

**Comment by dwgray** [comment]
@VividLemon This is the first component I've written from scratch, so I'd appreciate any early feedback you have on this before I document it and write tests (also interested in what you'd test on this - this is basically a very paired-down version of `BFormSelect` and that doesn't have direct testing - most of the complex code is shared code that is under test).

BTW - I'm not planning on filling in significant sized holes as I'm going through the parity/documentation pass - I think it's more important that I get through a first round. However, this helper component is actually something that I use and is blocking my next round of page conversions - it just happened to be related to stuff that I'm touching with the parity work.

**Comment by dwgray** [comment]
> @VividLemon This is the first component I've written from scratch...

This is ready to go as far as I'm concerned. But as I noted, this is my first component from scratch...

File: packages/bootstrap-vue-next/src/composables/useFormSelect.ts (hunk 1 above)
Comment (VividLemon) [inline packages/bootstrap-vue-next/src/composables/useFormSelect.ts:28]: `(unknown | Record<string, unknown>)[] = get(`
This is bad code on my part, as type 'unknown' already is inclusive of `Record<string, unknown>`

Comment (VividLemon) [reply packages/bootstrap-vue-next/src/composables/useFormSelect.ts:28]: Seemingly the best option would be to pass it in the parent component. 

`useFormSelect(() => props.options, computed(() => (...props, optionsField: props.optionsField ?? 'options'}))`

etc etc

Comment (VividLemon) [reply packages/bootstrap-vue-next/src/composables/useFormSelect.ts:28]: With this change this file would be pretty unchanged

File: packages/bootstrap-vue-next/src/components/BForm/BFormDatalist.vue [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+  disabled: false,
+  disabledField: 'disabled',
+  htmlField: 'html',
+  id: undefined,
+  // eslint-disable-next-line vue/require-valid-default-prop
+  options: () => [],
+  textField: 'text',
+  valueField: 'value',
+})
+const props = useDefaults(_props, 'BFormSelect')
```
Comment (VividLemon) [inline packages/bootstrap-vue-next/src/components/BForm/BFormDatalist.vue:33]: This seems like it needs a different key `BFormSelect`

File: packages/bootstrap-vue-next/src/components/BForm/BFormDatalist.vue (hunk 3 above)
Comment (VividLemon) [reply packages/bootstrap-vue-next/src/components/BForm/BFormDatalist.vue:50]: You'd have to show some code for what you tried.

I don't like this code here. 

I think if this component never has a situation where isComplex should be true, then it should never be an issue and this code should be removed. Ie if it does ever happen somehow, we should simply ignore it as its user error

Hypothetically this is something that would be validated by TS. Don't know how strongly typed that is though

**Review by VividLemon — CHANGES_REQUESTED** [review]

File: packages/bootstrap-vue-next/src/components/BForm/BFormDatalist.vue (hunk 3 above)
Comment (dwgray) [reply packages/bootstrap-vue-next/src/components/BForm/BFormDatalist.vue:50]: Here's a fiddle https://jsfiddle.net/ohdwg/6tzu4nqc/2/ showing that options groups aren't supported in data lists.  They're also not documented on Mozilla for datalist: https://jsfiddle.net/ohdwg/6tzu4nqc/2/ while they are for  select: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/select

I agree that typing the property would be the preferred way to go. But, the types have to be loose enough to allow substituting key names for the different fields (diabledField, htmlField, textField, valueField, labelField, optionsField).

I reworked this to silently ignore complex options and addressed the limitation in the documentation. I considered flattening the list (which is closer to what the native control does if you add in complex options), but then I'd have to add back in `labelField` and `optionsField` which seems like ugly complexity for something that is getting flattened back out. Short of pulling off some typescript voodoo, that's the best I can come up with.

File: packages/bootstrap-vue-next/src/composables/useFormSelect.ts (hunk 1 above)
Comment (dwgray) [reply packages/bootstrap-vue-next/src/composables/useFormSelect.ts:28]: Thanks - I like that better.

File: packages/bootstrap-vue-next/src/components/BForm/BFormDatalist.vue (hunk 4 above)
Comment (dwgray) [reply packages/bootstrap-vue-next/src/components/BForm/BFormDatalist.vue:33]: Fixed.

File: packages/bootstrap-vue-next/src/composables/useFormSelect.ts (hunk 1 above)
Comment (dwgray) [reply packages/bootstrap-vue-next/src/composables/useFormSelect.ts:28]: Sorry - yeah, I can change it to just unknown if you like. I can also take one more run at trying to more strictly type this - but I'd like you take on whether you think it's even doable before I spend more time on it (with the issue of the field names being fungible at runtime that I noted below).

Comment (dwgray) [reply packages/bootstrap-vue-next/src/composables/useFormSelect.ts:28]: On closer examination, I believe `undefined | unknown[]` is a slight improvement to this declaration. I think that's the best we can do in that case.

File: packages/bootstrap-vue-next/src/components/BForm/BFormDatalist.vue (hunk 4 above)
Comment (VividLemon) [reply packages/bootstrap-vue-next/src/components/BForm/BFormDatalist.vue:33]: No, I mean it's currently BFormSelect, but seems to need `BFormDatalist`

Comment (dwgray) [reply packages/bootstrap-vue-next/src/components/BForm/BFormDatalist.vue]: Sorry - something I don't understand is going on with github's view of this. If you look at [this commit](https://github.com/bootstrap-vue-next/bootstrap-vue-next/pull/2036/commits/9145882f360bfaed693326c4098f22221930668a) - I've changed line #33 of BFormDatlist.vue to 
`const props = useDefaults(_props, 'BFormDatalist')`, which is what you want, right? This comment is marked as outdated, so it's not updating the diff here? Not sure why.

File: packages/bootstrap-vue-next/src/components/BForm/BFormDatalist.vue [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+  default?: (props: Record<string, never>) => any
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  first?: (props: Record<string, never>) => any
+}>()
+
+const computedId = useId(() => props.id, 'datalist')
+
+const {normalizedOptions, isComplex} = useFormSelect(
+  () => props.options,
+  computed(() => ({...props, optionsField: 'options', labelField: 'label'})).value
```
Comment (VividLemon) [inline packages/bootstrap-vue-next/src/components/BForm/BFormDatalist.vue:46]: This is wrong. You should be passing in a Ref, not its .value =>

ie. if props changes, useFormSelect wont update

```suggestion
  computed(() => ({...props, optionsField: 'options', labelField: 'label'}))
```

**Review by VividLemon — APPROVED** [review]

File: packages/bootstrap-vue-next/src/composables/useFormSelect.ts [hunk 6]
```diff
@@ -25,7 +25,7 @@ export default (options: MaybeRefOrGetter, props: Record<string, unknown>) => {
     const html: string = get(option, props.htmlField as string)
     const disabled: boolean = get(option, props.disabledField as string)
 
-    const opts: undefined | (unknown | Record<string, unknown>)[] = props.optionsField
+    const opts: undefined | unknown[] = props.optionsField
```
Comment (VividLemon) [inline packages/bootstrap-vue-next/src/composables/useFormSelect.ts:28]: Some of the code for useFormSelect needs to be rewritten to support reactive inputs

This function
image
shadows this parameter
image
(Two duplicate variable names)

```ts
export default (options: MaybeRefOrGetter, props: Record<string, unknown>) => {
```
Options needs a generic input `MaybeRefOrGetter<unknown>` would suffice, unless there's some more strict input we can use

props should be `props: MaybeRefOrGetter<Record<string, unknown>>` which will likely need to be satisfied with

```ts
 const propsValue = readonly(toRef(props))
```

or

```ts
  const normalizeOption = (
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    option: any
  ): ComplexSelectOptionRaw | SelectOption => {
    const propsValue = toValue(props) // <-- then use this in the function
    if (typeof option === 'string') {
```

I prefer the first though. Perhaps `normalizeOptions(toValue(options)))` should be

```ts
const optionsValue = readonly(toRef(options))
const normalizedOptions = computed(() => normalizeOptions(optionsValue.value))
```
Here as well though

**Review by VividLemon — CHANGES_REQUESTED** [review]

File: packages/bootstrap-vue-next/src/composables/useFormSelect.ts [hunk 7]
```diff
@@ -1,15 +1,20 @@
...
+) => {
   const isComplex = (option: unknown): option is ComplexSelectOptionRaw =>
     typeof option === 'object' && option !== null && 'label' in option
 
   const normalizeOption = (
     // eslint-disable-next-line @typescript-eslint/no-explicit-any
     option: any
   ): ComplexSelectOptionRaw | SelectOption => {
+    const propsValue = readonly(toRef(props)) // as Readonly<Record<string, unknown>>
+
```
Comment (VividLemon) [inline packages/bootstrap-vue-next/src/composables/useFormSelect.ts:17]:
```suggestion
export default (
  options: MaybeRefOrGetter<Readonly<unknown[]>>,
  props: MaybeRefOrGetter<Record<string, unknown>>
) => {
  const propsValue = readonly(toRef(props)) // as Readonly<Record<string, unknown>>
  const optionsValue = readonly(toRef(options))

  const isComplex = (option: unknown): option is ComplexSelectOptionRaw =>
    typeof option === 'object' && option !== null && 'label' in option

  const normalizeOption = (
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    option: any
  ): ComplexSelectOptionRaw | SelectOption => {
```

File: packages/bootstrap-vue-next/src/composables/useFormSelect.ts [hunk 8]
```diff
@@ -45,11 +50,12 @@ export default (options: MaybeRefOrGetter, props: Record<string, unknown>) => {
...
 
   const normalizeOptions = (
-    options: (unknown | Record<string, unknown>)[]
+    opts: Readonly<unknown[]>
     // eslint-disable-next-line @typescript-eslint/no-explicit-any
-  ): (ComplexSelectOptionRaw | SelectOption)[] => options.map((option) => normalizeOption(option))
+  ): (ComplexSelectOptionRaw | SelectOption)[] => opts.map((option) => normalizeOption(option))
 
-  const normalizedOptions = computed(() => normalizeOptions(toValue(options)))
+  const optionsValue = readonly(toRef(options))
```
Comment (VividLemon) [inline packages/bootstrap-vue-next/src/composables/useFormSelect.ts:57]:
```suggestion
```

**Review by VividLemon — CHANGES_REQUESTED** [review]

File: packages/bootstrap-vue-next/src/composables/useFormSelect.ts [hunk 9]
```diff
@@ -1,8 +1,14 @@
-import {computed, type MaybeRefOrGetter, toValue} from 'vue'
+import {computed, type MaybeRefOrGetter, readonly, toRef} from 'vue'
 import {get} from '../utils'
 import type {ComplexSelectOptionRaw, SelectOption} from '../types'
 
-export default (options: MaybeRefOrGetter, props: Record<string, unknown>) => {
+export default (
+  options: MaybeRefOrGetter<Readonly<unknown[]>>,
```
Comment (VividLemon) [inline packages/bootstrap-vue-next/src/composables/useFormSelect.ts:6]:
```suggestion
  options: MaybeRefOrGetter<unknown[]>,
```

Comment (VividLemon) [inline packages/bootstrap-vue-next/src/composables/useFormSelect.ts:6]:
```suggestion
  options: MaybeRefOrGetter<ReadonlyArray<unknown>>,
```

File: packages/bootstrap-vue-next/src/composables/useFormSelect.ts [hunk 10]
```diff
@@ -45,11 +51,11 @@ export default (options: MaybeRefOrGetter, props: Record<string, unknown>) => {
   }
 
   const normalizeOptions = (
-    options: (unknown | Record<string, unknown>)[]
+    opts: Readonly<unknown[]>
```
Comment (VividLemon) [inline packages/bootstrap-vue-next/src/composables/useFormSelect.ts:54]:
```suggestion
    opts: ReadonlyArray<unknown>
```

**Review by VividLemon — APPROVED** [review]

**Pull request merged**